specification/redisenterprise/resource-manager/Microsoft.Cache/stable/2023-11-01/redisenterprise.json

```diff
@@ -0,0 +1,1954 @@
+      "description": "Describes the RedisEnterprise cluster",
+      "type": "object",
+      "properties": {
+        "sku": {
+          "$ref": "#/definitions/Sku",
+          "description": "The SKU to create, which affects price, performance, and features."
+        },
+        "zones": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+          "description": "The Availability Zones where this cluster will be deployed."
+        },
+        "identity": {
+          "$ref": "../../../../../common-types/resource-management/v4/managedidentity.json#/definitions/ManagedServiceIdentity",
+          "description": "The identity of the resource."
+        },
+        "properties": {
+          "x-ms-client-flatten": true,
+          "$ref": "#/definitions/ClusterProperties",
+          "description": "Other properties of the cluster."
+        }
+      },
+      "required": [
+        "sku"
+      ],
+      "allOf": [
+        {
+          "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/TrackedResource"
+        }
+      ]
+    },
+    "ClusterUpdate": {
+      "description": "A partial update to the RedisEnterprise cluster",
+      "type": "object",
+      "properties": {
+        "sku": {
+          "$ref": "#/definitions/Sku",
+          "description": "The SKU to create, which affects price, performance, and features."
+        },
+        "properties": {
+          "x-ms-client-flatten": true,
+          "$ref": "#/definitions/ClusterProperties",
+          "description": "Other properties of the cluster."
+        },
+        "identity": {
+          "$ref": "../../../../../common-types/resource-management/v4/managedidentity.json#/definitions/ManagedServiceIdentity",
+          "description": "The identity of the resource."
+        },
+        "tags": {
+          "type": "object",
+          "additionalProperties": {
+            "type": "string"
+          },
+          "x-ms-mutability": [
+            "read",
+            "create",
+            "update"
+          ],
+          "description": "Resource tags."
+        }
+      }
+    },
+    "ClusterProperties": {
+      "title": "RedisEnterprise cluster properties",
+      "description": "Properties of RedisEnterprise clusters, as opposed to general resource properties like location, tags",
+      "type": "object",
+      "properties": {
+        "minimumTlsVersion": {
+          "type": "string",
+          "description": "The minimum TLS version for the cluster to support, e.g. '1.2'",
+          "enum": [
+            "1.0",
+            "1.1",
+            "1.2"
+          ],
+          "x-ms-enum": {
+            "name": "TlsVersion",
+            "modelAsString": true
+          }
+        },
+        "encryption": {
+          "type": "object",
+          "description": "Encryption-at-rest configuration for the cluster.",
+          "properties": {
+            "customerManagedKeyEncryption": {
```

Comment:
>     "customerManagedKeyEncryption": { [](http://example.com/codeflow?start=7&length=38)

use the common definition here [https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/customermanagedkeys.json](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/customermanagedkeys.json)

Reply:
I think using the common-types would unfortunately make for a worse spec than avoiding it.

E.g. the API currently does not use many properties
```json
        "infrastructureEncryption": {
          "type": "string",
          "enum": [
            "enabled",
            "disabled"
          ],
          "x-ms-enum": {
            "name": "InfrastructureEncryption",
            "modelAsString": true
          },
          "description": "(Optional) Discouraged to include in resource definition. Only needed where it is possible to disable platform (AKA infrastructure) encryption. Azure SQL TDE is an example of this. Values are enabled and disabled."
        },
```

```json
                "federatedClientId": {
                  "type": "string",
                  "format": "uuid",
                  "description": "application client identity to use for accessing key encryption key Url in a different tenant. Ex: f83c6b1b-4d34-47e4-bb34-9d83df58b540"
                },
                "delegatedIdentityClientId": {
                  "type": "string",
                  "format": "uuid",
                  "description": "delegated identity to use for accessing key encryption key Url. Ex: /subscriptions/fa5fc227-a624-475e-b696-cdd604c735bc/resourceGroups/<resource group>/providers/Microsoft.ManagedIdentity/userAssignedIdentities/myId. Mutually exclusive with identityType systemAssignedIdentity and userAssignedIdentity - internal use only."
                }
```

Also, the identity type in common types is apparently not `x-ms-enum`?

```json
                "identityType": {
                  "type": "string",
                  "enum": [
                    "systemAssignedIdentity",
                    "userAssignedIdentity",
                    "delegatedResourceIdentity"
                  ],
                  "description": "The type of identity to use. Values can be systemAssignedIdentity, userAssignedIdentity, or delegatedResourceIdentity."
                },
```

Also, it seems valuable to have the custom comments in the custom definition. E.g. common-types states:

```
"description": "key encryption key Url, versioned or unversioned. Ex: https://contosovault.vault.azure.net/keys/contosokek/562a4bb76b524a1493a6afe8e536ee78 or https://contosovault.vault.azure.net/keys/contosokek."
```

Whereas for our service, due to a restriction imposed by the backing resource, only versioned keys are usable. So the custom type definition states:

```
"description": "Key encryption key Url, versioned only. Ex: https://contosovault.vault.azure.net/keys/contosokek/562a4bb76b524a1493a6afe8e536ee78"
```

Reply:
Sounds like a pretty good argument overall to me.

I suppose the x-ms-enum thing should be fixed in common-types though.